CIPs/0017.md

```diff
@@ -0,0 +1,96 @@
+```
+
+The `les` client must then be modified to send a `GetCertifiedEpochsValidatorSet` message during `lightest` sync. This will likely replace the function `getEpochOrNormalHeaders` within `eth/downloader`.
+
+Upon receiving the `CertifiedEpochsValidatorSet` response message, the light client must then verify headers and proofs provided. Proofs will be verified via a call into `celo-bls-go`'s `verify` fn.
+
+The verify function requires information on an epoch's validator set, so this will likely require some access changes (private -> public) and exposing more data structures to allow the validator set to be queried and modified upon proof verification.
+
+### Upgradeability
+In order to support an upgradeable snark circuit, we utilize a `versionNumber` parameter with every proof. This corresponds to a version of the snark circuit and thus a corresponding verification key. These version numbers will be stored on chain, within the `blockchain_parameters` smart contract, with the corresponding hash of the verification key. The clients are expected to store all applicable verification keys.
+
+For any uses of the snark's `verify` fn in the above protocols, a client is expected to:
+* Grab the verification key corresponding to the proof's `versionNumber`
+* Hash this key and compare the result to `blockchain_parameters` published VK hash.
+* If equal, the client may proceed to verify the proof.
+
+This allows the snark circuit to be upgraded via incrementing future proofs' `versionNumber` and publishing a new mapping of `versionNumber: VK_hash`, as well as distributing the new VK.
+
+### Bad proofs
+Bad proofs should be prevented from propagated throughout the network via the `verify` check that each node performs. In the case that any participant receives a proof that does not pass verification for its inputs, that participant is expected to drop the sender. 
+
+## Alternative Solutions
+
+Alternatively, `GetPlumoProofMsg` could utilize a bloom filter instead, allowing for a constant sized message. However, bloom filters allow for false positives: a false positive here would indicate to a peer that the sender has a proof when the sender does not. This could prevent certain proofs from ever getting to some nodes. For example, say node A is on the network and has received a certain amount of proofs via traditional propagation. Node A then goes down for a short period, and within that period proof P is added and propagated. Node A then comes back online, and sends a `GetPlumoProofsMsg` to request any proofs it may have missed, and sends along the corresponding bloom filter for the proofs it holds. If proof P happens to be a false positive for the set of proofs node A already has, then all of node A's peers will simply return nothing. Thus node A will never receive proof P. This can then propagate similarly through the network either through a similar path as node A, or for any nodes where all of their peers have already "lost" proof P.
```

Comment:
👍 One place where bloom filters could be handy is if the proof serving peer provides a filter to help answer the client peer's question of "does this peer have a proof with identifier X". That is not how the protocol is specified here though, so I agree that it doesn't seem relevant here.